homeassistant/components/shelly/sensor.py

```diff
@@ -255,30 +261,46 @@ async def async_setup_entry(
 class ShellySensor(ShellyBlockAttributeEntity, SensorEntity):
     """Represent a shelly sensor."""
 
+    def __init__(
+        self,
+        wrapper: ShellyDeviceWrapper,
+        block: aioshelly.Block,
+        attribute: str,
+        description: BlockAttributeDescription,
+    ) -> None:
+        """Initialize sensor."""
+        super().__init__(wrapper, block, attribute, description)
+        self._last_value: float | None = None
+
+        if description.last_reset == LAST_RESET_NEVER:
+            self._attr_last_reset = dt.utc_from_timestamp(0)
+        elif description.last_reset == LAST_RESET_UPTIME:
+            self._attr_last_reset = (
+                dt.utcnow() - timedelta(seconds=wrapper.device.status["uptime"])
+            ).replace(second=0, microsecond=0)
+
     @property
     def state(self) -> StateType:
         """Return value of sensor."""
+        if (
+            self.description.last_reset == LAST_RESET_UPTIME
+            and self.attribute_value is not None
+        ):
+            value = cast(float, self.attribute_value)
+
+            if self._last_value and self._last_value > value:
+                self._attr_last_reset = dt.utcnow().replace(second=0, microsecond=0)
+                _LOGGER.info("Energy reset detected for entity %s", self.name)
```

Comment:
We don't want side effects like logging in entity state properties. Please remove or move to a callback.